### DISCUSSION TITLE
How to determine a atom is primitive Atom?

### DISCUSSION THREAD

**Discussion by himself65** [pmndrs/jotai#1881] [Q&A] [answered]
```ts
function isPrimitiveAtom(atom: 
    Atom<unknown> |
    PrimitiveAtom<unknown> |
    WritableAtom<unknown, unknown[], unknown>
): atom is PrimitiveAtom<unknown> {
  // todo
}
```

**Comment by dai-shi** [comment]
PrimitiveAtom is a special type of WritableAtom.

https://github.com/pmndrs/jotai/blob/536dd3b2d9f42f7cdc7721d76b3a6e3b13d0e29f/src/vanilla/atom.ts#L55-L59

In this case, there's no way to distinguish.

If you want to determine if an atom is created with `atom(value)`
https://github.com/pmndrs/jotai/blob/536dd3b2d9f42f7cdc7721d76b3a6e3b13d0e29f/src/vanilla/atom.ts#L78-L81

then, you can do this hack.
https://github.com/pmndrs/jotai/blob/536dd3b2d9f42f7cdc7721d76b3a6e3b13d0e29f/src/vanilla/store.ts#L11-L14

**Reply by himself65** [reply]
Our app wants to distinguish the `primitive value,` `effect,` and `dispatch.` So that we can collect the side effect in server-side rendering

```ts
const primitiveAtom = atom('value')
const effectAtom = atom(async (get) => {
  return fetch(get(primitiveAtom)).then(res => res.json())
})
const dispatchAtom = atom(null, (get, set) => // ...)
```

**Reply by dai-shi** [reply, marked as answer]
It's out of the scope of the library. You can do anything on your end.
```js
// using a map
const atomTypes = new Map();

const primitiveAtom = atom('value');
atomTypes.set(primitiveAtom, 'primitive');
```

```js
// using a custom property
const primitiveAtom = atom('value');
primitiveAtom.atomType = 'primitive';
```
